ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/util/P2PDumpManager.java

```diff
@@ -0,0 +1,108 @@
+
+  private static final String GOSSIP_DECODING_ERROR_DIR = "gossip_decoding_error_messages";
+  private static final String GOSSIP_REJECTED_DIR = "rejected_gossip_messages";
+  private static final String INVALID_BLOCK_DIR = "invalid_blocks";
+
+  private final Path directory;
+
+  public P2PDumpManager(final Path directory) {
+    this.directory = directory;
+    if (this.directory.resolve(GOSSIP_DECODING_ERROR_DIR).toFile().mkdir()) {
+      LOG.info(
+          String.format(
+              "%s directory has been created to save gossip messages with decoding errors.",
+              GOSSIP_DECODING_ERROR_DIR));
+    }
+    if (this.directory.resolve(GOSSIP_REJECTED_DIR).toFile().mkdir()) {
+      LOG.info(
+          String.format(
+              "%s directory has been created to save rejected gossip messages.",
+              GOSSIP_REJECTED_DIR));
+    }
+    if (this.directory.resolve(INVALID_BLOCK_DIR).toFile().mkdir()) {
+      LOG.info(
+          String.format(
+              "%s directory has been created to save invalid blocks.", INVALID_BLOCK_DIR));
+    }
+  }
+
+  public String saveGossipMessageDecodingError(
+      final String topic, final String arrivalTimestamp, final Bytes originalMessage) {
+    final String fileName = String.format("%s_%s.ssz", arrivalTimestamp, topic);
+    final String identifiers = String.format("Topic: %s", topic);
+    return saveBytesToFile(
+        "gossip message with decoding error",
+        GOSSIP_DECODING_ERROR_DIR,
+        fileName,
+        identifiers,
+        originalMessage);
+  }
+
+  public String saveGossipRejectedMessageToFile(
+      final String topic, final String arrivalTimestamp, final Bytes decodedMessage) {
+    final String fileName = String.format("%s_%s.ssz", arrivalTimestamp, topic);
+    final String identifiers = String.format("Topic: %s", topic);
+    return saveBytesToFile(
+        "rejected gossip message", GOSSIP_REJECTED_DIR, fileName, identifiers, decodedMessage);
+  }
+
+  public String saveInvalidBlockToFile(
+      final UInt64 slot, final Bytes32 blockRoot, final Bytes blockSsz) {
+    final String fileName =
+        String.format("slot%s_root%s.ssz", slot, blockRoot.toUnprefixedHexString());
+    final String identifiers = String.format("Slot: %s, Block Root: %s", slot, blockRoot);
+    return saveBytesToFile("invalid block", INVALID_BLOCK_DIR, fileName, identifiers, blockSsz);
+  }
+
+  private String saveBytesToFile(
+      final String object,
+      final String errorDirectory,
+      final String fileName,
+      final String identifiers,
+      final Bytes bytes) {
+    final Path path = directory.resolve(errorDirectory).resolve(fileName);
+    try {
+      // Create file and save ssz
+      if (!path.toFile().createNewFile()) {
+        final String errorMessage =
+            String.format("Unable to create new file to save %s. %s", object, identifiers);
+        throw new FileAlreadyExistsException(errorMessage);
```

Comment:
We don't even want to throw an exception from here without handling it. Imagine when we call this from `ForkChoice`, and then if this exception is thrown, we have no control on how this is going to impact the flow of the code.

Whatever happens, we want to make sure it gets handled within `P2PDumpManager` and does not impact the "outside world".